app.js

```diff
@@ -1,146 +1,90 @@
-import countries from './countries'
+import countries from './countries';
 
-const team1Element = document.querySelector('#team1')
-const team2Element = document.querySelector('#team2')
-const team1img = document.querySelector('.teaminner1')
-const team2img = document.querySelector('.teaminner2')
-const textzone = document.querySelector('#score')
+const MAIN_LOOP_INTERVAL = 5000;
+const MATCH_LOOP_INTERVAL = 20000;
 
-var oldmatchid = null
+const team1 = {
+	score: document.getElementById('team1'),
+	img: document.querySelector('.teaminner1')
+};
+const team2 = {
+	score: document.getElementById('team2'),
+	img: document.querySelector('.teaminner2')
+};
+const textzone = document.querySelector('#score');
 
-getjson()
-var jsoninterval = setInterval(function () {
-	getjson()
-}, 5000)
+let previousMatch = null;
+let matchInterval = null;
 
-function getjson() {
-	fetch('/settings')
-		.then((res) => res.json())
-		.then((settingsdata) => {
-			let { apikey, matchid, warmups, reverse, bestof } = settingsdata
-
-			if (oldmatchid != matchid) {
-				oldmatchid = matchid
-				if (apikey && matchid && warmups && reverse && bestof) {
-					matchdata(apikey, matchid, warmups, osuinterval, reverse, bestof)
-					var osuinterval = setInterval(function () {
-						matchdata(apikey, matchid, warmups, osuinterval, reverse, bestof)
-					}, 20000)
-				}
-			}
-		})
+async function main() {
+	const { apiKey, matchId, warmups, reverse, bestOf } = await fetch('/settings').then(res => res.json());
+	if (apiKey === undefined ||
```

Comment:
the condition can be cleaner and also there must be an implementation for the `""` string and `"undefined"` string cases. It should look like this: `(!apiKey || apiKey == '' || apiKey == 'undefined' || !matchId || matchId == '' || matchId == 'undefined' || !warmups || !reverse || !bestOf)` (00:06 AM code quality)

Reply:
Why though? At first I also wanted to do something like `!apiKey || !matchId || !warmups || !reverse || !bestOf` but realized that we don't want to check if the values are falsy but if they're *defined*. IMO since we're running the backend locally anyway we should do validation before sending saved settings to the server **and** have proper error handling in case the API key is wrong or something, not validate the data in `app.js`.

Reply:
so we have to do the string check at the settings page, right?

Reply:
I think that'd be the best and the simplest for us, yeah